src/main/java/com/google/sps/data/User.java

```diff
@@ -223,7 +251,8 @@ public Entity toEntity() {
       userEntity.setProperty(ADDED_TO_CALENDAR_EVENTS, this.added_to_calendar_events);
       userEntity.setProperty(CURRENT_CHALLENGE, this.current_challenge_id);
       userEntity.setProperty(CHALLENGE_STATUSES, embedChallengeStatuses());
-      userEntity.setProperty(COMPLETED_CHALLENGES,  getCompletedChallengesAsArray());
+      userEntity.setProperty(COMPLETED_CHALLENGES,  convertHashSetToArrayList(this.completed_challenges));
+      userEntity.setProperty(EARNED_BADGES, convertHashSetToArrayList(this.earned_badges));
       return userEntity;
   }
 
```

Comment:
Won't let me add a comment on the actual function but can you add checks for your added fields (completed challenges & earned badges) to the equals method on line 266?